### PR TITLE
Add write function to deviceshifumqtt.

### PR DESCRIPTION
**Pull request by hdy8200**
**What this PR does / why we need it**:
deviceshifumqtt didn't have a write function before.

**Which issue this PR fixes**

 fixes # Add write function to deviceshifumqtt.

**How is this PR tested**

- [x] unit test
- [ ] e2e test
- [ ] other (please specify)

### CODE REVIEW

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go [hunk 1]
```diff
@@ -31,6 +32,8 @@ const (
 )
 
 var (
+	client                      mqtt.Client
+	MQTTTopic 					string
```
Comment (BtXin) [inline pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go:36]: Format this line please.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go [hunk 2]
```diff
@@ -5,3 +5,9 @@ type ReturnBody struct {
 	MQTTMessage   string `json:"mqtt_message"`
 	MQTTTimestamp string `json:"mqtt_receive_timestamp"`
 }
+
+// RequestBody Body of mqtt's request by POST method
+type RequestBody struct {
+	MQTTTopic 	  string `json:"mqtt_topic"`
+	MQTTMessage   []byte `json:"mqtt_message"`
```
Comment (BtXin) [inline pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go:12]: format.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go [hunk 3]
```diff
@@ -5,3 +5,9 @@ type ReturnBody struct {
 	MQTTMessage   string `json:"mqtt_message"`
 	MQTTTimestamp string `json:"mqtt_receive_timestamp"`
 }
+
+// RequestBody Body of mqtt's request by POST method
+type RequestBody struct {
+	MQTTTopic 	  string `json:"mqtt_topic"`
```
Comment (BtXin) [inline pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go:11]: We defined which mqttTopic to read from edge_device `CRD`, and for which mqttTopic to write to, I think we also need to define it in edge_device `CRD` for consistent.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go (hunk 1 above)
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go]: This has been modified.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go (hunk 2 above)
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go:12]: This has been modified.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go (hunk 3 above)
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go]: I have removed this line.

**Review by tomqin93 — CHANGES_REQUESTED** [review]
@hdy8200 added some comment, please take a look

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go [hunk 4]
```diff
@@ -5,3 +5,8 @@ type ReturnBody struct {
 	MQTTMessage   string `json:"mqtt_message"`
 	MQTTTimestamp string `json:"mqtt_receive_timestamp"`
 }
+
+// RequestBody Body of mqtt's request by POST method
+type RequestBody struct {
+	MQTTMessage    []byte `json:"mqtt_message"`
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go:11]: let's use string

**Review by tomqin93 — CHANGES_REQUESTED** [review]
@hdy8200 added another comment, please take a look

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go [hunk 5]
```diff
@@ -143,8 +146,33 @@ func (handler DeviceCommandHandlerMQTT) commandHandleFunc() http.HandlerFunc {
...
+		} else if reqType == http.MethodPost {
+			body, err := io.ReadAll(r.Body)
+			if err != nil {
+				klog.Errorf("Error when Read Data From Body, error: %v", err)
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+
+			requestBody := RequestBody{}
+			err = json.Unmarshal(body, &requestBody)
```
[the remark this answers is not in the document]
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go]: I have modified.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go [hunk 6]
```diff
@@ -143,8 +146,33 @@ func (handler DeviceCommandHandlerMQTT) commandHandleFunc() http.HandlerFunc {
...
+			if err != nil {
+				klog.Errorf("Error to Unmarshal request body to struct")
+				http.Error(w, "unexpected end of JSON input", http.StatusBadRequest)
+				return
+			}
+			klog.Infof("requestBody: %v", requestBody)
+
+			token := client.Publish(MQTTTopic, 1, false, requestBody.MQTTMessage)
+			if token.Error() != nil {
```
[the remark this answers is not in the document]
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt.go:162]: I have added.

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt_test.go [hunk 7]
```diff
@@ -112,9 +156,44 @@ func TestCommandHandleMQTTFunc(t *testing.T) {
...
+	client = mqtt.NewClient(opts)
+
+	MQTTTopic = "/test/test"
+	requestBody := RequestBody{
+		MQTTMessage: []byte("1234"),
+	}
 
 	// test post method
-	r := dc.Post().Do(context.TODO())
+	// test post method when MQTTServer not connected
```
[the remark this answers is not in the document]
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt_test.go:170]: I have removed

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go (hunk 4 above)
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqttconfig.go]: I have modified

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt_test.go [hunk 8]
```diff
@@ -36,6 +57,28 @@ func TestMain(m *testing.M) {
 	if err != nil {
 		klog.Fatal(err)
 	}
+
+	stop <- struct{}{}
+	wg.Wait()
+	os.Unsetenv("SERVER_LISTEN_PORT")
```
Comment (tomqin93) [inline pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt_test.go:63]: can we use similar approach as PR https://github.com/Edgenesis/shifu/pull/375

**Review by tomqin93 — COMMENTED** [review]
@hdy8200 added 1 comment, the rest looks good

File: pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt_test.go (hunk 8 above)
Comment (hdy8200) [reply pkg/deviceshifu/deviceshifumqtt/deviceshifumqtt_test.go]: Nice!  I have already used it

**Review by tomqin93 — APPROVED** [review]
LGTM

**Review by BtXin — APPROVED** [review]

**Pull request merged**